src/main/java/org/opentripplanner/routing/impl/OrcaFareServiceImpl.java

```diff
@@ -0,0 +1,404 @@
+                // The start of a free transfer must be with a transit agency that permits it!
+                freeTransferStartTime = ride.startTime;
+            }
+            float singleLegPrice = getRidePrice(ride, fareType, fareRules);
+            float discountedFare = getDiscountedFare(fareType, rideType, singleLegPrice, ride.routeData);
+            if (hasFreeTransfers(fareType, rideType) && inFreeTransferWindow(freeTransferStartTime, ride.startTime)) {
+                // If using Orca (free transfers), the total fare should be equivalent to the
+                // most expensive leg of the journey.
+                orcaFareDiscount = Float.max(orcaFareDiscount, discountedFare);
+            } else if (hasAgencyDiscount(rideType)) {
+                // If not using Orca and an agency discount is available, add this to the overall cost.
+                cost += discountedFare;
+            } else {
+                // If free transfers are not permitted and there are no agency specific discounts, add the cash price of
+                // this leg to the total cost.
+                cost += singleLegPrice;
+            }
+            if (!inFreeTransferWindow(freeTransferStartTime, ride.startTime)) {
+                // If the trip time has exceeded the free transfer time limit of two hours the rider is required to
+                // purchase a new fare. This also resets the free transfer trip window, applies the orca discount to the
+                // overall cost and then resets the orca fare discount ready for the next transfer window.
+                freeTransferStartTime = ridePermitsFreeTransfers ? ride.startTime : null;
+                cost += orcaFareDiscount;
+                orcaFareDiscount = 0;
+            }
+        }
+        cost += orcaFareDiscount;
+        if (cost < Float.POSITIVE_INFINITY) {
+            fare.addFare(fareType, getMoney(currency, cost));
+        }
+        return cost > 0 && cost < Float.POSITIVE_INFINITY;
+    }
+
+    /**
+     *  Trip within the free two hour transfer window.
+     */
+    private boolean inFreeTransferWindow(Long freeTransferStartTime, long currentLegStartTime) {
+        return freeTransferStartTime != null &&
+            currentLegStartTime < freeTransferStartTime + FREE_TRANSFER_TIME_DURATION;
+    }
+
+    /**
+     * A free transfer can be applied if using Orca and the transit agency permits free transfers.
+     */
+    private boolean hasFreeTransfers(Fare.FareType fareType, RideType rideType) {
+        return permitsFreeTransfers(rideType) && usesOrca(fareType);
+    }
+
+    /**
+     * Has agency specific discount. This is separate to Orca.
+     */
+    private boolean hasAgencyDiscount(RideType rideType) {
+        return rideType == RideType.WASHINGTON_STATE_FERRIES ||
+               rideType == RideType.KITSAP_TRANSIT_FAST_FERRY_WESTBOUND ||
+               rideType == RideType.KITSAP_TRANSIT_FAST_FERRY_EASTBOUND ;
```

Comment:
I believe the Kitsap fast allow transfers, so I think this line should just be:

```suggestion
               rideType == RideType.SKAGIT_TRANSIT;
```

Reply:
I think this should say as is. The method `permitsFreeTransfers` handles this and includes `WASHINGTON_STATE_FERRIES` and `SKAGIT_TRANSIT`. The method `hasAgencyDiscount` defines agencies which have discount separate to Orca.

Reply:
Yes, I saw that logic. So in this following code, here's what'll happen:

```java
if (hasFreeTransfers(fareType, rideType) && inFreeTransferWindow(freeTransferStartTime, ride.startTime)) {
    // If using Orca (free transfers), the total fare should be equivalent to the
    // most expensive leg of the journey.
    orcaFareDiscount = Float.max(orcaFareDiscount, discountedFare);
} else if (hasAgencyDiscount(rideType)) {
    // If not using Orca and an agency discount is available, add this to the overall cost.
    cost += discountedFare;
} else {
    // If free transfers are not permitted and there are no agency specific discounts, add the cash price of.   
    // this leg to the total cost.
    cost += singleLegPrice;
}
```

When Skagit Transit is encountered, `hasFreeTransfers` will always evaluate to false because of the logic you mention in `permitsFreeTransfers`, therefore the next conditional will be analyzed. At this point, I would expect `hasAgencyDiscount ` to say that Skagit Transit has an agency discount. It is a little confusing to me because basically every agency has a discounted fare, but since my recommendation above would be to include Skagit along with WSF, this would then ensure that those agencies that also happen to not allow free transfers are able to add on the discounted fare if it applies. Actually, after looking at this further wouldn't this add on the `singleLegPrice` for an Orca agency after the free transfer window expires in addition to the `orcaFareDiscount`?

Reply:
@evansiroky I've updated the logic around this which should have simplified things.